src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs

```diff
@@ -101,20 +101,41 @@ static string DecodeFrom64(string encodedData)
         /// <returns>Subscription addresses that were found for the provided messageTypes</returns>
         public async Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context)
         {
-            var subscribers = new List<Subscriber>();
+            var subscribers = new HashSet<Subscriber>(new SubscriberComparer());
             var table = client.GetTableReference(subscriptionTableName);
 
             foreach (var messageType in messageTypes)
             {
-                var query = new TableQuery<Subscription>().Where(TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, messageType.ToString()));
+                var name = messageType.TypeName;
+                var lastNameChar = name[name.Length - 1];
+                var nextChar = (char)(lastNameChar + 1);
+                var lowerBound = TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.GreaterThanOrEqual, name);
+                var upperBound = TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.LessThan, name.Substring(0, name.Length - 1) + nextChar);
+                var query = new TableQuery<Subscription>().Where(TableQuery.CombineFilters(lowerBound, "and", upperBound));
 
                 var subscriptions = await table.ExecuteQueryAsync(query).ConfigureAwait(false);
                 var results = subscriptions.Select(s => new Subscriber(DecodeFrom64(s.RowKey), s.EndpointName));
 
-                subscribers.AddRange(results);
+                foreach (var subscriber in results)
+                {
+                    subscribers.Add(subscriber);
+                }
             }
 
             return subscribers;
         }
+
+        class SubscriberComparer : IEqualityComparer<Subscriber>
+        {
+            public bool Equals(Subscriber x, Subscriber y)
+            {
+                return StringComparer.InvariantCulture.Compare(x.Endpoint, y.Endpoint) == 0 && StringComparer.InvariantCulture.Compare(x.TransportAddress, y.TransportAddress) == 0;
```

Comment:
This could be, but it would change the behavior as two types could use upper/lower letters and before the fix would be considered not equal (different partition keys).